### PR TITLE
chore: refactor and splitting the code

### PR DESCRIPTION
**Pull request by jobo322**

### CODE REVIEW

File: src/ngmca.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  updateLambda,
+  updateMatrixA,
+  updateMatrixS,
+} from './stages';
+
+/**
+ * Performing non-negative matrix factorization solving argmin_(A >= 0, S >= 0) 1 / 2 * ||Y - AS||_2^2 + lambda * ||S||_1
+ * @param {*} originalMatrix - Matrix to be separated.
+ * @param {Number} rank - The maximum number of linearly independent column/row vectors in the matrix.
+ * @param {Object} options - Options of ngmca factorization method.
```
Comment (lpatiny) [inline src/ngmca.js]: [options={}]

File: src/ngmca.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  initialize,
+  normalize,
+  updateLambda,
+  updateMatrixA,
+  updateMatrixS,
+} from './stages';
+
+/**
+ * Performing non-negative matrix factorization solving argmin_(A >= 0, S >= 0) 1 / 2 * ||Y - AS||_2^2 + lambda * ||S||_1
+ * @param {*} originalMatrix - Matrix to be separated.
```
Comment (lpatiny) [inline src/ngmca.js]: Should not be '*' what it may be ? Matrix||Array[][] ?

File: src/ngmca.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+ * @param {Number} [options.maximumIteration = 500] - Maximum number of iterations.
+ * @param {Number} [options.maxFBIteration = 80] - Maximum number of iterations of the Forward-Backward subroutine.
+ * @param {Object} [options.randGenerator = Math.random] - Random number generator for the subroutine of initialization.
+ * @param {Number} [options.maxInitFBIteration = 50] - Maximum number of iterations of the Forward-Backward subroutine at the initialization.
+ * @param {Number} [options.toleranceFB = 1e-5] - relative difference tolerance for convergence of the Forward-Backward sub-iterations.
+ * @param {Number} [options.toleranceFBInit = 0] - relative difference tolerance for convergence of the Forward-Backward sub-iterations at the initialization.
+ * @param {Number} [options.phaseRatio = 0.8] - transition between decreasing thresholding phase and refinement phase in percent of the iterations.
+ * @param {Number} [options.tauMAD = 1] - constant coefficient for the final threshold computation.
+ * @param {Boolean} [options.useTranspose = false] - if true the originalMatrix is transposed.
+ */
```
Comment (lpatiny) [inline src/ngmca.js:24]: What does it return ?

File: src/util/index.js [hunk 4]
```diff
@@ -0,0 +1,6 @@
+export * from './zeroInsteadOfNegative';
+export * from './normBy';
+export * from './normProj';
+export * from './getMax';
+export * from './getMedians';
+export * from './dimMADstd';
```
Comment (lpatiny) [inline src/util/index.js]: This file is useless. You may directly import the right util method that you need

File: src/stages/index.js [hunk 5]
```diff
@@ -0,0 +1,6 @@
+export * from './checkMatrixS';
+export * from './initialize';
+export * from './updateLambda';
+export * from './updateMatrixS';
+export * from './updateMatrixA';
+export * from './normalize';
```
Comment (lpatiny) [inline src/stages/index.js]: This file is useless. You may directly import the right util method that you need

**Review by lpatiny — COMMENTED** [review]
Is it possible to add some basic jsdoc to all the 'internal' methods ?

**Comment by lpatiny** [comment]
Please could you also improve the README ?

I would like to know how to use this package in a real situation and what is the actual result.

I have a combined vector [1,2,3,4]

and pure components:
- [1,1,0,0]
- [0,1,1,1]
- [0,0,1,1]
- [0,0,0,1]

How to get the answer

I have a more than one combined vector
`[
  [1,2,3,4],
  [1,2,3,0]
]`
Same pure components. How to get the results ?

I think the method could either accept a vector or a matrix

File: src/ngmca.js (hunk 2 above)
Comment (jobo322) [reply src/ngmca.js]: Matrix||Array<Array> is it ok?

**Pull request merged**